CARDS-1474: Associate an optional frequency with PROMs questionnaires

acrowthe opened the pull request:
Contains subtasks CARDS-1482, CARDS-1483 and CARDS-1488. CARDS-1487 is still outstanding.

marta- committed:
CARDS-1474/CARDS-1482: Update questionnaire set to specify frequencies

acrowthe committed:
CARDS-1483: Create empty forms based on visit requirements
- Add frequency fields to PROMs questionairres
- In progress: Create ResourceChangeListener that triggers on Form changes
  - Current status: triggers on all forms in /libs/

acrowthe committed:
CARDS-1483: Create empty forms based on visit requirements
When a Visit Information form is created, create all forms associated
with the specified questionnaire set that are in need of filling out per
the questionnaire set's frequency

acrowthe committed:
CARDS-1488: Update surveys_complete when visit forms are complete
- Skip visits that do not have any forms or that only have a
  Visit Information form

sashaandjic commented:
build is failing
<img width="670" alt="image" src="https://user-images.githubusercontent.com/23243244/146631857-33e5241e-4cd3-4fb8-830a-f3de97314f9c.png">

acrowthe committed:
CARDS-1488: Update surveys_complete when visit forms are complete
Fix checkstyle error

acrowthe commented:
> build is failing
Oops, comment javadoc was in the wrong order. Quick fix is up, should be building now

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/EQ5D.xml:
```diff
@@ -42,6 +42,11 @@
 		<value>Health  Questionnaire</value>
 		<type>String</type>
 	</property>
+	<property>
+		<name>frequency</name>
+		<value>2</value>
+		<type>Long</type>
+	</property>
```
The frequency should not be defined in the questionnaire node, as different clinics may want different frequencies for the same questionnaire. The frequency should be only in the [QuestionnaireSet](https://github.com/data-team-uhn/cards/pull/839/files#diff-dd5d57fb44be9906a1ca60a394452347aa4eeb4f5f53b97a348ec9d4c2982425R41-R45) instead.

acrowthe committed:
CARDS-1483: Create empty forms based on visit requirements
- Remove frequency from questionnaires in favor of questionnaire set frequency
- Fix statusFlags being a single value string rather than multivalue
  - Fixes subject view crash
- Ignore incomplete forms when comparing frequencies

acrowthe commented:
Issue to fix:

Upon checking a visit information form:
- If no forms are created and the visit does not have any forms already created, mark the form as completed
- If there are any other visits within the past 72 hours, do not duplicate their forms regardless of completion status.

acrowthe committed:
Cards-1474: Associate an optional frequency with PROMs questionnaires
- Mark forms as complete when they have no forms that need to be completed
- Do not duplicate forms between vists within 72 hours, regardless of completion
  - Forms should be completed at the visit if not completed before

sdumitriu commented on proms-resources/backend/pom.xml:
```diff
@@ -0,0 +1,76 @@
+    <dependency>
+      <groupId>io.uhndata.cards</groupId>
+      <artifactId>cards-dataentry</artifactId>
+      <version>0.9-SNAPSHOT</version>
```
```suggestion
      <version>${project.version}</version>
```

sdumitriu commented on proms-resources/backend/pom.xml:
```diff
@@ -0,0 +1,76 @@
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <scope>provided</scope>
```
This is already the default, specified in the parent pom.

```suggestion
```

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/nodetypes/proms.cnd:
```diff
@@ -53,6 +53,9 @@
   // An optional estimate for the time it would take to fill out the questionnaire, in minutes
   - estimate (long)
 
+  // An optional frequency for re-taking the questionnaire, in weeks
```
I wonder if this should be in days instead, we may have other projects in the future with more frequent visits. @marta- ?

sdumitriu commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -24,13 +24,25 @@
+      "service.ranking:Integer":200,
+      "scripts":["create service user proms-visit-backend \n set ACL on /Forms,/Questionnaires,/Subjects,/SubjectTypes \n   allow jcr:read for proms-visit-backend \n end \n set ACL on /Forms \n   allow jcr:read,rep:write,jcr:versionManagement for proms-visit-backend \n end \n"]
+    },
+    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~":{
```
This needs a specific name:

```suggestion
    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~proms":{
```

sdumitriu commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -24,13 +24,25 @@
+    },
+    "org.apache.sling.jcr.repoinit.RepositoryInitializer~proms_backend":{
+      "service.ranking:Integer":200,
+      "scripts":["create service user proms-visit-backend \n set ACL on /Forms,/Questionnaires,/Subjects,/SubjectTypes \n   allow jcr:read for proms-visit-backend \n end \n set ACL on /Forms \n   allow jcr:read,rep:write,jcr:versionManagement for proms-visit-backend \n end \n"]
```
Why does this need to be in a separate configuration? Can it be all in one?

```suggestion
        "create service user proms-visit-backend \n set ACL on /Forms,/Questionnaires,/Subjects,/SubjectTypes \n   allow jcr:read for proms-visit-backend \n end \n set ACL on /Forms \n   allow jcr:read,rep:write,jcr:versionManagement for proms-visit-backend \n end \n",
        "create user patient \n set ACL for patient \n     allow jcr:read on /Questionnaires,/SubjectTypes \n     deny jcr:all on /Questionnaires restriction(rep:itemNames,audit_intro,audit_results,eq5d_results,gad7_intro,gad7_results,phq9_intro,phq9_results) \n     allow jcr:read,jcr:write,jcr:nodeTypeManagement,jcr:versionManagement on /Forms restriction(cards:sessionSubject) \n     allow jcr:read on /Subjects restriction(cards:sessionSubject) \n     deny jcr:modifyProperties on /Forms restriction(rep:itemNames,subject,questionnaire) \n end"
      ]
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,643 @@
+    public void handleEvent(Event event)
+    {
+        try (ResourceResolver localResolver = this.resolverFactory.getServiceResourceResolver(null)) {
+            this.resolver = localResolver;
```
This is a singleton, but it could handle multiple events in parallel. Setting these as instance fields will likely lead to concurrency issues, where a thread handling an event may suddenly work with a different session than the one it created, or it may end up using a closed session.

There are three workarounds for this:
- set them as `ThreadLocal` members instead (easy, minor performance hit)
- don't have any state, always pass down the session object to inner functions (easy, slightly uglier code)
- use a factory for the actual service registration, creating new event handlers whenever an event is created (harder and not that useful)

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,643 @@
+    @Override
+    public void handleEvent(Event event)
+    {
+        try (ResourceResolver localResolver = this.resolverFactory.getServiceResourceResolver(null)) {
```
Looks like there will be multiple services in this module, including the Torch import that Francis is working on, and the patient authentication validation that I am working on. I would use a subservice for each of these, to make permissions more fine-grained and explicit.

sdumitriu committed:
[cleanup] Applied codestyle

sdumitriu committed:
CARDS-1474 / CARDS-1483: When a new visit is created for a patient, automatically create empty forms for the questionnaires that need to be filled out for that visit
Bugfix: singleton services should not hold state

sashaandjic reviewed (approved):
tested

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,643 @@
+     * @param days number of days being added
+     * @return the resulting calenar
+     */
+    private static Calendar addDays(Calendar date, int days)
```
I wonder if we should ignore the actual time, i.e. is a visit 3 days and 2 hours earlier considered in the range of the current visit or not?

sdumitriu committed:
[misc] Added more utilities for working with data

sdumitriu committed:
CARDS-1474 / CARDS-1483: When a new visit is created for a patient, automatically create empty forms for the questionnaires that need to be filled out for that visit
Replace the OSGi's event listener mechanism with Sling's resource change listener, since the former sometimes fails to work

sdumitriu committed:
CARDS-1474 / CARDS-1483: When a new visit is created for a patient, automatically create empty forms for the questionnaires that need to be filled out for that visit
Various changes:
- use the data-entry utilities in favor of custom code for accessing data
- use a subservice for the service authentication
- bugfix: save/checkout race condition with the original change that triggered the event listener
- codestyle improvements

acrowthe commented:
"The patient completed the pre-appointment surveys" is not updated from true to false when new forms are created for an existing visit. In the reproduction steps, do not use the save button: instead, just save the forms automatically by navigating off the page.
1. Create visit 1, with a visit info form, date and surveys.
2. Create visit 2, with a date 2 days after visit 1, and valid surveys. Observe visit 2 is complete as no surveys were created
3. Delete visit 1
4. Edit visit 2's date and leave the page to autosave. Observe visit 2 now has a full complement of surveys, but is still marked as complete

Possible bug. Shouldn't ever occur, but would break the whole listener for a specific patient if it does:
If a visit information form is deleted or otherwise inaccessible without deleting other forms for that visit, all future visit change listener calls to handleVisitInformationForm crash due to a NPE in `isWithinDateRange`
1. Create visit information form with Date & surveys
2. Delete the previous visit information form
3. Try to create another visit information form for a different visit with Date and surveys
4. Observe that no forms were made for the new visit and there is a NPE warning in error.log

Not really a bug, but the log gets spammed with errors when deleting a visit subject which has a child visit information form:
`17.01.2022 16:52:01.399 *ERROR* [oak-executor-112] io.uhndata.cards.proms.internal.VisitChangeListener Node with path /Forms/4e007527-8ee5-420e-b88a-9e1037ae8d0c does not exist.
javax.jcr.PathNotFoundException: Node with path /Forms/4e007527-8ee5-420e-b88a-9e1037ae8d0c does not exist.`
It appears that 1 instance of this message appears any time a form is deleted (visit information, EQ5D, PHQ, etc.)

acrowthe commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,613 @@
+                    // All done, exit the try-loop
+                    break;
+                } catch (final RepositoryException e) {
+                    LOGGER.error("Failed to checkin form: {}", e.getMessage(), e);
```
Should this be a warning instead of an error on the first 2 attempts, if it is somewhat expected and recovered from? Adding "Trying again" could make the message clearer too

sdumitriu replied:
Indeed.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,613 @@
+        }
+
+        final Calendar triggeringDate = visitInformation.getVisitDate();
+        final boolean isWithinVisitMargin = isWithinDateRange(visitDate, triggeringDate, VISIT_CREATION_MARGIN_DAYS);
```
This will fail if the visit does not have a Visit Information form, or if no date is specified in it. Needs a null check.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -0,0 +1,613 @@
+                    visitDate = (Calendar) this.formUtils
+                        .getValue(this.formUtils.getAnswer(form, visitInformation.getVisitDateQuestion()));
+                } else {
+                    questionnaires.put(questionnaire.getIdentifier(), !isFormIncomplete(form));
```
In Cardio, all the questionnaires can be completed only once, but to be safe, I would change this to record if at least one form of each type is complete:

```suggestion
                    questionnaires.merge(questionnaire.getIdentifier(), !isFormIncomplete(form), Boolean::logicalOr);
```

sdumitriu committed:
[misc] Comment improvements

sdumitriu committed:
CARDS-1474 / CARDS-1483: When a new visit is created for a patient, automatically create empty forms for the questionnaires that need to be filled out for that visit
- bugfix: deleting a form triggers an exception
- bugfix: a visit without a visit date (or no visit information form at all) causes exceptions when creating another visit
- compact the feature configuration
- use $project.version for inter-module dependencies
- support more than one form of a type per visit

sdumitriu commented:
> "The patient completed the pre-appointment surveys" is not updated from true to false when new forms are created for an existing visit.

True, but this isn't a targeted scenario. It is assumed that all the forms are pre-created, and once the patient has filled them all in, it's done-done, no further forms are to be added to the visit. As far as the patient is concerned, he saw the "all done, thank you and buh-bye" message, he's not going to go back and check if there's something new to fill in.

sashaandjic commented:
build fails
[CARDS-1474.txt](https://github.com/data-team-uhn/cards/files/7885311/CARDS-1474.txt)

marta- commented:
> build fails [CARDS-1474.txt](https://github.com/data-team-uhn/cards/files/7885311/CARDS-1474.txt)

The build works for me and I don't see anything in the latest code that could cause that error. However, there was a forced push a couple of hours ago, so you might be having an old version of the branch or a merge conflict.

sashaandjic commented:
> The build works for me and I don't see anything in the latest code that could cause that error. However, there was a forced push a couple of hours ago, so you might be having an old version of the branch or a merge conflict.

I closed all my running instances and made a new clone. It builds with no problem now.

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/149972816-5f9ab868-4c36-4de4-8572-c5b7e1961a8f.png)

acrowthe commented:
Tested, looks good

sashaandjic commented:
Should we be generating surveys if there is no Visit info form for the given visit?

marta- commented:
> Should we be generating surveys if there is no Visit info form for the given visit?

We shouldn't, are we?

sashaandjic commented:
> We shouldn't, are we?

yes. if you do not import csv files and login as one of the test users the survey forms will be generated even though there is no Visit Info form.

sashaandjic commented:
QUESTION: Patient fills in all surveys and we set `The patient completed the pre-appointment surveys` invthe Visit Info form to `Yes`. Physician then adds a new form. The Visit is still marked as completed even though patient can now fill out the newly added survey for the same visit.
Should we upon form creation reset the flag `The patient completed the pre-appointment surveys`?

sashaandjic commented:
created first Visit info for for October 2021; got all 5 forms created; filled in all surveys and had my form marked as surveys completed
created another Visit 3 moths later; this time I got only 3 surveys forms created as expected; when I signed in as patient I was told that I already submitted all surveys. We should switch from visit 1 to visit 2 here.
created another Visit few weeks later and no surveys forms were created (correct)

sashaandjic reviewed (approved):
tested
